pkg/handlers/launch.go

```diff
@@ -144,23 +144,119 @@ type launchHandler struct {
 	if slackClient == nil {
 		return nil, errors.New("unexpected state. Slack client is nil")
 	}
+	ctx, cancel := context.WithCancel(context.Background())
+
+	h := &launchHandler{
+		client:               client,
+		kubeClient:           kubeClient,
+		slackClient:          slackClient,
+		lastFailureTime:      make(map[string]time.Time),
+		sleep:                time.Sleep,
+		processToWaitFor:     make(chan k6.TestRun, 1),
+		waitForProcessesDone: make(chan struct{}, 1),
+		ctx:                  ctx,
+	}
+	h.cancelWaitForProcesses = cancel
+	go h.waitForProcesses(ctx, maxProcessHandlers)
+	return h, nil
+}
+
+func (h *launchHandler) Close() {
+	if h.cancelWaitForProcesses != nil {
+		h.cancelWaitForProcesses()
+	}
+	<-h.waitForProcessesDone
+}
+
+// context exposes the internal context for testing purposes.
+//
+//nolint:unused
+func (h *launchHandler) context() context.Context {
+	return h.ctx
+}
+
+// waitForProcesses handles incoming processes and waits for them to complete.
+// This way we can avoid k6 jobs where we do not need the results to become
+// zombie processes.
+func (h *launchHandler) waitForProcesses(ctx context.Context, maxProcessHandlers int) {
+	defer func() {
+		h.waitForProcessesDone <- struct{}{}
+	}()
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		<-ctx.Done()
+		close(h.processToWaitFor)
+		wg.Done()
+	}()
+	// We have a fixed number of available process handlers:
+	availableProcessHandlers := make(chan struct{}, maxProcessHandlers)
+	for i := 0; i < maxProcessHandlers; i++ {
+		availableProcessHandlers <- struct{}{}
+	}
+loop:
+	for {
+		select {
+		case cmd := <-h.processToWaitFor:
+			// If we have a handler available, we can proceed:
+			log.Debugf("waiting for an available process handler")
+			select {
+			case <-ctx.Done():
+				break loop
+			case <-availableProcessHandlers:
+				log.Debugf("process handler available")
+				wg.Add(1)
+				go func() {
+					h.waitForProcess(cmd)
+					availableProcessHandlers <- struct{}{}
+					wg.Done()
+				}()
+			}
+		case <-ctx.Done():
+			break loop
+		}
+	}
+	wg.Wait()
+	close(availableProcessHandlers)
+}
+
+func (h *launchHandler) waitForProcess(cmd k6.TestRun) {
+	if cmd == nil {
+		log.Warnf("nil as testrun passed")
+		return
+	}
+	pid := cmd.PID()
+	log.WithField("pid", pid).Debug("waiting for testrun to exit")
+	_ = cmd.Wait()
+	log.WithField("pid", pid).Debugf("testrun exited")
+}
 
-	return &launchHandler{
-		client:          client,
-		kubeClient:      kubeClient,
-		slackClient:     slackClient,
-		lastFailureTime: make(map[string]time.Time),
-		sleep:           time.Sleep,
-	}, nil
+// registerProcessCleanup adds a handler to the process so that it will
+// eventually be closed and its resources returned.
+//
+// Note that this method can actually block which will, in turn, cause the
+// calling HTTP handler to be blocked.
```

Comment:
hmm, that sounds like it might introduce regressions if we have people using this to launch a lot of tests. they might start seeing timeouts?

anything we can do about that? send a 429 back?

Reply:
The problem right now is that the test is actually already running at this point. One thing I could do is basically request a token from the process-handler *before* the actual test starts to run. Not sure if its worth the effort, though.

Reply:
Done. I've also exposed the number of still available concurrent tests as a metric.